### PR TITLE
BUG: Series/Index arithmetic result names with NAs

### LINKED ISSUE #22041: Pandas Series: unexpected error while doing difference (`_maybe_match_name`)
pandas version is 0.22.0, numpy version is 1.14.0

While performing the following:

```python
a = pd.Series([1, 2, 3])
a.name = np.int64(123)

b = pd.Series([4, 5, 6])
b.name = (123, 789)

a - b
```
I obtain the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-63-e4efbf202d3c> in <module>()
----> 1 one - two

~/venv/lib/python3.4/site-packages/pandas/core/ops.py in wrapper(left, right, name, na_op)
    727 
    728         if isinstance(rvalues, ABCSeries):
--> 729             name = _maybe_match_name(left, rvalues)
    730             lvalues = getattr(lvalues, 'values', lvalues)
    731             rvalues = getattr(rvalues, 'values', rvalues)

~/venv/lib/python3.4/site-packages/pandas/core/common.py in _maybe_match_name(a, b)
    137     b_has = hasattr(b, 'name')
    138     if a_has and b_has:
--> 139         if a.name == b.name:
    140             return a.name
    141         else:

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### LINKED ISSUE #39757: ValueError from operations between Series with specific names
The following raises a ValueError due to comparison of the two names in the function _maybe_match_name (core/ops/common.py).

```python
>>> from pandas import Series
>>> import numpy as np
>>> series_a = Series([1,2,3], name=np.float64(1.0))
>>> series_b = Series([1,2,3], name=(np.float64(1.0), np.float64(2.0)))
>>> series_a + series_b
```

It seems to only occur when comparison a tuple of numpy dtypes to a single numpy dtype, but I didn't test every possible combination. It does not occur when using built-in floats, I imagine this is due to overloading of the == operator by the numpy dtypes when the other operand is a tuple.

There is a nearby TODO "what if they both have np.nan for their names?" I don't know how feasible it is to consider every case where equality fails, but it seems a try/except returning None would seem to resolve this issue, and be in line with the default return value if the names do not compare equal. I would be keen to implement this simple fix if it seems sensible.

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.0.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252
pandas           : 1.0.3
numpy            : 1.19.0
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.4
setuptools       : 47.3.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.0
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: 0.8.1
bs4              : 4.9.1
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : 3.2.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : 0.51.2

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #39757
- [x] closes #22041
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/ops/common.py [hunk 1]
```diff
@@ -116,10 +117,21 @@ def _maybe_match_name(a, b):
     a_has = hasattr(a, "name")
     b_has = hasattr(b, "name")
     if a_has and b_has:
-        if a.name == b.name:
-            return a.name
-        else:
-            # TODO: what if they both have np.nan for their names?
+        try:
+            if a.name == b.name:
```
Comment (jreback) [inline pandas/core/ops/common.py:121]: are we somehow raising on NA comparing to a string?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops/common.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/ops/common.py:121]: if either is pd.NA then this raises TypeError

**Comment by jreback** [comment]
looks fine can you add a whatsnew note

**Comment by jbrockmendel** [comment]
whatsnew added + greenish

**Review by jreback — APPROVED** [review]

**Comment by rhshadrach** [comment]
Certainly no objection to the modification here, but I do wonder if we should be changing the API to narrow what is allowed as a Series name or DataFrame label. For example, from #39757,
```
series_a = Series([1,2,3], name=np.float64(1.0))
series_b = Series([1,2,3], name=(np.float64(1.0), np.float64(2.0)))
pd.concat([series_a, series_b], axis=1).columns.sort_values()
```

still fails. I am wondering if we should change the API to only allow basic primitive types (Boolean, int, str, float?, perhaps some others) and tuples of primitive types.

**Comment by jreback** [comment]
the problem is that the name can be any pandas primitive or tuplr if those (as we allow this for index labels) which doesn't it much

**Comment by rhshadrach** [comment]
@jreback - I think you're saying that even if we are add some restrictions, `np.float64` (and hence, tuples of such objects) should be allowed as a name. Is that right?

**Comment by jreback** [comment]
yes ultimately we allow these as names (well have to be hashable) or we could inhibit a number of downstream ops

**Pull request merged**